### PR TITLE
[timer] multiple microseconds timer support

### LINKED ISSUE #1945: Multiple microseconds timer support
Current OT only support multiple milliseconds timer, and a single microseconds timer used by [CSMA  backoff](https://github.com/openthread/openthread/blob/master/src/core/mac/mac.cpp#L109) (only enabled by nRF52840 now).

To support the coming new features, we need multiple microseconds timer, the change proposal:

1. Platform alarm APIs update

    - Combine current two alarm APIs ([alarm](https://github.com/openthread/openthread/blob/master/include/openthread/platform/alarm.h), [usec-alarm](https://github.com/openthread/openthread/blob/master/include/openthread/platform/usec-alarm.h)) to single microseconds alarm API, platforms which could not support microseconds can still provide milliseconds accuracy but with the same microseconds alarm APIs.

2. `class Timer` update
    Option 1:
    - Replace current `uint32_t  mFireTime` field with the structure [otPlatUsecAlarmTime](https://github.com/openthread/openthread/blob/master/include/openthread/platform/usec-alarm.h#L61)
    - This option looks much concise, although it will increase 2 bytes RAM space for each `Timer`. 

    Option 2:
    - Create a new `class TimerUsec` based on `class Timer`, with an extra microseconds field.
    - This option will complicate the `class TimerScheduler` to handle the two timers (two queues, time comparison between them), it will increase some code size.

I prefer option 1 more, please let me know your preference and other better options.

### PR DESCRIPTION
**Pull request by chshu**
Resolves #1945.

* Remove TimerSchedulerLocator, Timer inherit from Ip6Locator instead to get TimerScheduler

* Add separate UsecTimer and UsecTimerScheduler for multiple microseconds timer support

### CODE REVIEW

**Comment by jwhui** [comment]
Thanks for this latest update - this is what I had in mind when I commented in #1945.

I know it will result in a bigger change, but I think it will be cleaner if the names make the resolution explicit.  For example:

- `TimerMilli`: the millisecond timer object representation
- `TimerMicro`: the microsecond timer object representation
- `TimerMilliScheduler`: the millisecond timer scheduler
- `TimerMicroScheduler`: the microsecond timer scheduler

Also, I think it will be cleaner if we rename the header files to:

- `include/openthread/platform/alarm-milli.h`
- `include/openthread/platform/alarm-micro.h`

Thoughts?

**Comment by chshu** [comment]
@jwhui, I have updated the names according to your suggestion, thanks.

@hubertmis, I also don't like the duplication codes:) So I tried with some other approach (as mentioned in #1945), but finally I find that the implementation in this PR seems to be most efficient (320 bytes flash and 32bytes RAM addition).

About the duplication codes, these functions take different Timer parameters and call into different alarm APIs, so a simple Base class could not work, we may need to introduce some other methods like Template, but it is not commonly used by OT. So if the flash and RAM addition is acceptable, I'd prefer to keep it as is (clean and readable).

**Comment by hubertmis** [comment]
@chshu I would prefer templates like in #1267 to avoid duplicated code. But it introduces maintenance cost (I'm not sure if cost of templates is greater than cost of code copy). Because of that I'm OK with current proposal as well.

**Comment by abtink** [comment]
@chshu first let me thank you for working on this, having flexible `TimerMilli` and `TimerMicro` will be super helpful. 

I have an idea for how to reuse the code for both types of timers. The key point as you mentioned is that the two types use different `Alarm` related APIs. What if we provide a pointer to the `Alarm` functions as an input parameter for methods in a common base class (virtual function trick without being virtual).

For example, we can define an `AlarmApi` struct like this:

````
struct AlarmApi
{
    void (*AlarmStartAt)(otInstance *aInstance, uint32_t aT0, uint32_t aDt);
    void (*AlarmStop)(otInstance *aInstance);
    uint32_t (*AlarmGetNow)(void);
};
````
Then defining a base Timer class like this (I removed all comments)

````
class TimerBase: public Ip6Locator, public Context
{
public:

    typedef void (*Handler)(Timer &aTimer);

    Timer(TimerScheduler &aScheduler, Handler aHandler, void *aContext):
        TimerSchedulerLocator(aScheduler),
        Context(aContext),
        mHandler(aHandler),
        mFireTime(0),
        mNext(this) {
    }

    uint32_t GetFireTime(void) const { return mFireTime; }

    bool IsRunning(void) const { return (mNext != this); }

    // `Start()`, `StartAt()`, `Stop()`, and `GetNow()` will be defined by sub-classes.

private:
    // The code for this can be shared, only change we may need is 
    // to pass the aNow as an input parameter (to avoid the need 
    // for having `GetNow()` method)
    bool DoesFireBefore(const Timer &aTimer, uint32_t aNow);

    void Fired(void) { mHandler(*this); }

    Handler         mHandler;
    uint32_t        mFireTime;
    Timer          *mNext;
};

class TimerMilli : public TimerBase
{
public:
    enum
    {
        kMaxDt = (1UL << 31) - 1,  //< Maximum permitted value for parameter `aDt` in `Start` and `StartAt` method.
    };

    // Here we define `Start` and `StartAt` and `Stop` 
    // and `GetNow` for the timer type using its 
    // corresponding scheduler.
}

// And same way for `TimerMicor` class.

````

We then have a `TimerSchedulerBase` which contains all the common code.

````
class TimerSchedulerBase: public Ip6Locator
{
    TimerSchedulerBase(void);

    void Add(TimerBase &aTimer, const AlarmApi &aAlarmApi);
    void Remove(TimerBase &aTimer, const AlarmApi &aAlarmApi);
    void ProcessTimer(const AlarmApi  &aAlarmApi);
private:
    void SetAlarm(const AlarmApi &aAlarmApi);
    static bool IsStrictlyBefore(uint32_t aTimeA, uint32_t aTimeB);

    TimerBase *mHead;
};
````

The common code use the input parameter `aAlarmApi` function pointers.

Then the two sub-classes for milli and micor scheduler will contain inline calls into base class providing the alarm API:

````
class TimerSchedulerMilli : public TimerSchedulerBase
{
public:
    TimerShceduelr(...)

    void Add(TimerMilli &aTimer)     { TimerSchedulerBase::Add(aTimer, sMilliAlarmApi); }
    void Remove(TimerMilli &aTimer)  { TimerSchedulerBase::Remove(aTimer, sMilliAlarmApi); }
    void ProcessTimer(void)  { TimerSchedulerBase::Process(aTimer, sMilliAlarmApi); }
    // ...

private:
    static const AlarmApi sMilliAlarmApi;
};

// In cpp file we define the `MilliAlarmApi` struct as a constant:

static const TimerSchedulerMilli::sMilliAlarmApi = { &otPlatAlarmStartAt, &otAlarmStop, &otAlarmgetNow };
````

And same way for the Micro one.
One thing I realized that we may need to make the `TimerScheduelrBase` be a `Ip6Locator` itself. For single-instance case this should not add any new variable and make the code simpler.

Thoughts?

(sorry for rather long comment... :)...)

**Comment by chshu** [comment]
@abtink , Thanks a lot for the detailed and great suggestion. I have submitted a new commit which implements the new Timer code structure. Please have a review again.

The new commit helps to reduce 72 bytes flash:

Base (Master):

```
$ arm-none-eabi-size output/nrf52840/bin/ot-cli-ftd 
   text	   data	    bss	    dec	    hex	filename
 158072	    524	  33374	 191970	  2ede2	output/nrf52840/bin/ot-cli-ftd
```

Before the new commit (304 bytes flash and 32 bytes RAM addition):
```
$ arm-none-eabi-size output/nrf52840/bin/ot-cli-ftd 
   text	   data	    bss	    dec	    hex	filename
 158376	    524	  33406	 192306	  2ef32	output/nrf52840/bin/ot-cli-ftd
```

After the new commit (232 bytes flash and 32 bytes RAM addition):
```
$ arm-none-eabi-size output/nrf52840/bin/ot-cli-ftd 
   text	   data	    bss	    dec	    hex	filename
 158304	    524	  33406	 192234	  2eeea	output/nrf52840/bin/ot-cli-ftd
```

File: src/core/common/timer.hpp [hunk 1]
```diff
@@ -62,56 +64,70 @@ class Timer;
  */
 
 /**
- * This class implements the timer scheduler.
+ * The Alarm APIs definition
  *
  */
-class TimerScheduler
+struct AlarmApi
```
Comment (abtink) [inline src/core/common/timer.hpp]: Suggestion: We can consider adding the `struct AlarmApi` type definition in the `TimerSchedulerBase` class (as a `protected` type). It hides the definition and made it only accessible by `TimerSchedulerBase` and its subclasses.

File: src/core/common/timer.hpp [hunk 2]
```diff
@@ -62,56 +64,70 @@ class Timer;
...
+
+/**
+ * This class implements the base timer scheduler.
+ *
+ */
+class TimerSchedulerBase: public Ip6Locator
+{
+    friend class TimerBase;
 
 public:
```
Comment (abtink) [inline src/core/common/timer.hpp]: How about if we make the constructors (and actually may be all the methods) in `TimerSchedulerBase` as `protected`? So only sub-classes can see/access them and  not allow instantiation of `TimerSchedulerBase` type object on its own.

File: src/core/common/timer.hpp [hunk 3]
```diff
@@ -128,16 +144,60 @@ class TimerScheduler
...
+
+    /**
+     * This method removes a timer instance to the timer scheduler.
+     *
+     * @param[in]  aTimer  A reference to the timer instance.
+     *
+     */
+    void Remove(TimerMilli &aTimer) { TimerSchedulerBase::Remove(reinterpret_cast<TimerBase &>(aTimer), sAlarmMilliApi); };
+
+
```
Comment (abtink) [inline src/core/common/timer.hpp]: nit: extra new line

File: src/core/common/timer.hpp [hunk 4]
```diff
@@ -128,16 +144,60 @@ class TimerScheduler
...
+        TimerSchedulerBase(aIp6) {
+    }
+
+    /**
+     * This method adds a timer instance to the timer scheduler.
+     *
+     * @param[in]  aTimer  A reference to the timer instance.
+     *
+     */
+    void Add(TimerMilli &aTimer) { TimerSchedulerBase::Add(reinterpret_cast<TimerBase &>(aTimer), sAlarmMilliApi); }
```
Comment (abtink) [inline src/core/common/timer.hpp]: I was thinking if we can avoid the `reinterpret_cast<>` here but may be tricky (as complier reaching here does not yet know `TimerMilli` inherits from `TimerBase`). We can consider changing the order of declaration of `TimerMilliShceduler` and `TimerMilli` classes. It then creates problem that the we cannot use `TimerMilliShceduler` methods in inline definitions of `StartAt()` and `Stop()` but I think we can move the definitions into the `.cpp` file.

**Comment by chshu** [comment]
@abtink, Thanks for the suggestions, I have updated the implementation accordingly.

File: src/core/common/timer.hpp [hunk 5]
```diff
@@ -62,56 +63,224 @@ class Timer;
...
+    Handler         mHandler;
+    uint32_t        mFireTime;
+    TimerBase      *mNext;
+};
+
+/**
+ * This class implements the millisecond timer.
+ *
+ */
+class TimerMilli: public TimerBase
```
Comment (jwhui) [inline src/core/common/timer.hpp]: We could probably rename `TimerBase` to just `Timer`, given that we already have more-specific `TimerMilli` and `TimerMicro`.

File: src/core/common/timer.hpp [hunk 6]
```diff
@@ -62,56 +63,224 @@ class Timer;
...
-    void SetAlarm(void);
+    TimerMilliScheduler &GetTimerMilliScheduler(void) const;
+};
 
+
+/**
+ * This class implements the base timer scheduler.
+ *
+ */
+class TimerSchedulerBase: public Ip6Locator
```
Comment (jwhui) [inline src/core/common/timer.hpp]: We could probably rename `TimerSchedulerBase` to `TimerScheduler`.

File: src/core/mac/mac.hpp [hunk 7]
```diff
@@ -691,11 +690,13 @@ class Mac: public ThreadNetifLocator
 
     static Mac &GetOwner(const Context &aContext);
 
-    Timer mMacTimer;
-#if !OPENTHREAD_CONFIG_ENABLE_PLATFORM_USEC_BACKOFF_TIMER
-    Timer mBackoffTimer;
+    TimerMilli mMacTimer;
+#if OPENTHREAD_CONFIG_ENABLE_PLATFORM_USEC_TIMER
+    TimerMicro mBackoffTimerMicro;
```
Comment (jwhui) [inline src/core/mac/mac.hpp]: I think it's fine to reuse `mBackoffTimer` here.

**Comment by chshu** [comment]
@jwhui , I have updated it according to your comments, thanks.

**Pull request merged**